toolchain/sem_ir/impl.h

```diff
@@ -50,20 +35,46 @@ struct Impl : public Printable<Impl> {
   InstId witness_id = InstId::Invalid;
 };
 
+// An implementation of a constraint. See EntityWithParamsBase regarding the
+// inheritance here.
+struct Impl : public EntityWithParamsBase,
+              public ImplFields,
+              public Printable<Impl> {
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "{self: " << self_id << ", constraint: " << constraint_id << "}";
+  }
+
+  // Determines whether this impl has been fully defined. This is false until we
+  // reach the `}` of the impl definition.
+  auto is_defined() const -> bool { return witness_id.is_valid(); }
+
+  // Determines whether this impl's definition has begun but not yet ended.
+  auto is_being_defined() const -> bool {
+    return definition_id.is_valid() && !is_defined();
+  }
+};
+
 // A collection of `Impl`s, which can be accessed by the self type and
 // constraint implemented.
 class ImplStore {
  public:
-  // Looks up the impl with this self type and constraint, or creates a new
-  // `Impl` if none exists.
-  // TODO: Handle parameters.
-  auto LookupOrAdd(TypeId self_id, TypeId constraint_id) -> ImplId {
-    auto result = lookup_.Insert(std::pair{self_id, constraint_id}, [&]() {
-      return values_.Add({.self_id = self_id, .constraint_id = constraint_id});
-    });
-    return result.value();
+  // TODO: Switch to something like TinyPtrVector. We expect it to be rare for
+  // there to be more than one ImplId per bucket.
+  using IdVector = llvm::SmallVector<ImplId, 1>;
+
+  // Returns the lookup bucket containing the list of impls with this self type
+  // and constraint, or adds a new bucket if this is the first time we've seen
+  // an impl of this kind. The lookup bucket only includes impls from the
+  // current file and its API file.
```

Comment:
```suggestion
  // an impl of this kind. Callers must only append impls from the
  // current file and its API file.
```

Reply:
Rephrased to clarify that this is a constraint on the caller.